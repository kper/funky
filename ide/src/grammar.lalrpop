use std::str::FromStr;
use crate::ir::ast::*;

grammar;

Comma<T>: Vec<T> = { 
    <v:(<T> ",")*> <e:T?> => match e {
        None => v,
        Some(e) => {
            let mut v = v;
            v.push(e);
            v
        }
    }
};

pub Program : Program = {
   <functions:Function*> => Program {
       functions: functions
   } 
};

pub Function : Function = {
    "define" <name:Id> "(result" <result:Num> ")" "(define" <def:Reg*> ")" "{" <instr:Instruction*>  "}" ";" => {
        Function {
            name,
            params: Vec::new(),
            definitions: def,
            results_len: result as usize,
            instructions: instr
        }
    },
    "define" <name:Id> "(param" <reg:Reg*> ")" "(result" <result:Num> ")" "(define" <def:Reg*> ")" "{" <instr:Instruction*>  "}" ";" => {
        Function {
            name,
            params: reg,
            definitions: def,
            results_len: result as usize,
            instructions: instr
        }
    },
    "define" <name:Name> "(result" <result:Num> ")" "(define" <def:Reg*> ")" "{" <instr:Instruction*>  "}" ";" => {
        Function {
            name,
            params: Vec::new(),
            definitions: def,
            results_len: result as usize,
            instructions: instr
        }
    },
    "define" <name:Name> "(param" <reg:Reg*> ")" "(result" <result:Num> ")" "(define" <def:Reg*> ")" "{" <instr:Instruction*>  "}" ";" => {
        Function {
            name,
            params: reg,
            definitions: def,
            results_len: result as usize,
            instructions: instr
        }
    }
};

pub Instruction : Instruction = {
    "BLOCK" <n:Id> => Instruction::Block(n),
    <dest:Reg> "=" <src:Reg> => Instruction::Assign(dest, src),
    <dest:Reg> "=" <n:Num> => Instruction::Const(dest, n),
    <dest:Reg> "=" "UNKNOWN" => Instruction::Unknown(dest),
    <dest:Reg> "=" "op" <src:Reg> => Instruction::Unop(dest, src),
    <dest:Reg> "=" <src1:Reg> "op" <src2:Reg> => Instruction::BinOp(dest, src1, src2),
    <dest:Reg> "=" "phi" <src1:Reg> <src2:Reg> => Instruction::Phi(dest, src1, src2),
    <dest:Reg> "=" "STORE" <src:Reg> "TO" <offset:Reg> "ALIGN" <align:Num> <size:Num> => Instruction::Store,
    "GOTO" <n:Id> => Instruction::Jump(n),
    "IF" <reg:Reg> "THEN" "GOTO" <n:Id> "ELSE" "GOTO" <n2:Id> => {
        Instruction::Conditional(reg, vec![n, n2])
    },
    "IF" <reg:Reg> "THEN" "GOTO" <n:Id> => {
        Instruction::Conditional(reg, vec![n])
    },
    "TABLE GOTO" <n:Id*> "ELSE" "GOTO" <n2:Id> => {
        let mut n = n.clone();
        n.push(n2);
        Instruction::Table(n)
    },
    "KILL" <reg:Reg> => Instruction::Kill(reg),
    "CALL" <n:Name> "(" <r:Reg*> ")" => Instruction::Call(n, r, vec![]),
    <dest:Reg*> "<-" "CALL" <n:Name> "(" <r:Reg*> ")" => Instruction::Call(n, r, dest),
    "CALL" <n:Id> "(" <r:Reg*> ")" => Instruction::Call(n, r, vec![]),
    <dest:Reg*> "<-" "CALL" <n:Id> "(" <r:Reg*> ")" => Instruction::Call(n, r, dest),
    "RETURN" <n:Reg*> ";" => {
        Instruction::Return(n)
    },
    "CALL INDIRECT" <n:Id*> "(" <p:Reg*> ")" => Instruction::CallIndirect(n, p, vec![]),
    <dest:Reg*> "<-" "CALL INDIRECT" <n:Id*> "(" <p:Reg*> ")" => Instruction::CallIndirect(n, p, dest),
};

Id : String = {
    <n:Num> => (n as usize).to_string()
};

pub Reg: String = {
    "%" <n:Num> => format!("%{}", n),
    "%" <n:Name> => format!("%{}", n),
};

pub Num: f64 = <s:r"[+-]?[0-9]*[.,]?[0-9]+"> => f64::from_str(s).unwrap();
pub Name: String = <s:r"[a-zA-Z]+"> => s.to_string();