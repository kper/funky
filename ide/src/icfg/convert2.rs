use crate::icfg::graph2::*;
use crate::ir::ast::Function as AstFunction;
use crate::ir::ast::Instruction;
/// This module is responsible to parse
/// the webassembly AST to a graph
use crate::{counter::Counter, solver::Request};
use anyhow::{bail, Context, Result};
use std::collections::VecDeque;

use log::debug;

use std::collections::HashMap;

use crate::ir::ast::Program;

use crate::icfg::graph2::Edge;

#[derive(Debug)]
pub struct ConvertSummary {
    block_counter: Counter,
}

impl ConvertSummary {
    pub fn new() -> Self {
        Self {
            block_counter: Counter::default(),
        }
    }

    pub fn visit(&mut self, prog: &Program, req: &Request) -> Result<Graph> {
        let mut graph = Graph::new();

        self.tabulate(&mut graph, prog, req)?;

        Ok(graph)
    }

    /// computes all intraprocedural edges
    fn flow(
        &self,
        function: &AstFunction,
        graph: &mut Graph,
        pc: usize,
        variable: &String,
    ) -> Result<Vec<Edge>> {
        debug!(
            "Calling flow for {} with var {} with pc {}",
            function.name, variable, pc
        );

        let mut edges = Vec::new();

        let instructions = &function.instructions;

        let instruction = instructions.get(pc).context("Cannot find instr")?;
        debug!("Next instruction is {:?}", instruction);

        //graph.add_statement(function, instruction, pc + 1)?;
        let before: Vec<Fact> = graph
            .get_facts_at(&function.name, pc)?
            .into_iter()
            .filter(|x| &x.belongs_to_var == variable)
            .cloned()
            .collect();

        debug!("Facts before statement {}", before.len());

        //graph.add_statement(function, format!("{:?} ({})", instruction, pc + 1), pc + 1)?;

        let after: Vec<_> = graph
            .get_facts_at(&function.name, pc + 1)?
            .into_iter()
            .filter(|x| &x.belongs_to_var == variable)
            .collect();

        debug!("Facts after statement {}", after.len());

        for (b, a) in before.into_iter().zip(after) {
            edges.push(Edge::Normal {
                from: b,
                to: a.clone().clone(),
                curved: false,
            });
        }

        Ok(edges)
    }

    /// Computes call-to-start edges
    fn pass_args(
        &self,
        program: &Program,
        caller_function: &AstFunction,
        callee: &String,
        params: &Vec<String>,
        dests: &Vec<String>,
        graph: &mut Graph,
        current_pc: usize,
    ) -> Result<Vec<Edge>> {
        let callee_function = program
            .functions
            .iter()
            .find(|x| &x.name == callee)
            .context("Cannot find function")?;

        let facts = graph.init_function(callee_function)?;
        let caller_facts = graph.get_facts_at(&caller_function.name, current_pc)?;

        let mut edges = Vec::new();
        for (caller, callee) in caller_facts.iter().zip(facts) {
            edges.push(Edge::Call {
                from: caller.clone().clone(),
                to: callee,
            })
        }

        Ok(edges)
    }

    /// Computes exit-to-return edges
    fn return_val(
        &self,
        caller_function: &String,
        callee_function: &String,
        caller_pc: usize,
        callee_pc: usize,
        //dest: &Vec<String>,
        caller_instructions: &Vec<Instruction>,
        graph: &mut Graph,
    ) -> Result<Vec<Edge>> {
        debug!("Trying to compute return_val");
        debug!("Caller: {} ({})", caller_function, caller_pc);
        debug!("Callee: {} ({})", callee_function, callee_pc);

        let dest = match caller_instructions.get(caller_pc).as_ref() {
            Some(Instruction::Call(_, _params, dest)) => dest,
            Some(x) => bail!("Wrong instruction passed to return val. Found {:?}", x),
            None => bail!("Cannot find instruction while trying to compute exit-to-return edges"),
        };

        let caller_facts = graph.get_facts_at(caller_function, caller_pc)?;
        let callee_facts = graph.get_facts_at(callee_function, callee_pc)?;

        debug!("caller_facts {:#?}", caller_facts);
        debug!("callee_facts {:#?}", callee_facts);

        // Generate edges when for all dest + taut

        let mut edges = Vec::new();

        // taut
        /*edges.push(Edge::CallToReturn {
            from: caller_facts.get(0).unwrap().clone().clone(),
            to: callee_facts.get(0).unwrap().clone().clone(),
        });*/

        debug!("=> dest {:?}", dest);

        /*
        for (i, d) in dest.iter().enumerate() {
            let caller_fact = caller_facts
                .iter()
                .find(|x| &x.belongs_to_var == d)
                .context("Cannot find var for caller_fact")?;
            let callee_fact = callee_facts
                .get(i + 1)
                .context("Cannot find var for callee_fact")?;

            edges.push(Edge::CallToReturn {
                from: caller_fact.clone().clone(),
                to: callee_fact.clone().clone(),
            });
        }*/

        let mut index = 0;
        for callee_fact in callee_facts.iter() {
            let caller_fact = caller_facts
                .get(index)
                .context("Cannot find var for caller_fact")?;

            if caller_fact.var_is_taut || dest.contains(&caller_fact.belongs_to_var) {
                edges.push(Edge::CallToReturn {
                    from: caller_fact.clone().clone(),
                    to: callee_fact.clone().clone(),
                });

                index += 1;
            }
        }

        Ok(edges)
    }

    /// Computes call-to-return
    fn call_flow(
        &self,
        program: &Program,
        caller_function: &AstFunction,
        callee: &String,
        params: &Vec<String>,
        dests: &Vec<String>,
        graph: &mut Graph,
        pc: usize,
    ) -> Result<Vec<Edge>> {
        debug!("Generating call-to-return edges for {}", callee);

        /* 
        graph.add_statement(
            caller_function,
            format!("{:?}", "call"),
            pc + 1,
            &"taut".to_string(),
        )?;*/
        //for dest in dests.iter() {
            //graph.add_statement(caller_function, format!("{:?}", "call"), pc)?;
            //graph.add_statement(caller_function, format!("{:?}", "call"), pc + 1)?;
        //}

        let before: Vec<_> = graph
            .get_facts_at(&caller_function.name, pc)?
            .into_iter()
            .map(|x| x.clone())
            .collect();
        debug!("Facts before statement {}", before.len());

        let after = graph.get_facts_at(&caller_function.name, pc + 1)?;

        debug!("Facts after statement {}", after.len());

        let after: Vec<_> = after
            .into_iter()
            .filter(|x| !dests.contains(&x.belongs_to_var))
            .collect();

        debug!("Facts after statement without dests {}", after.len());

        debug!("before {:#?}", before);
        debug!("after {:#?}", after);

        let mut edges = Vec::with_capacity(after.len());
        for fact in after {
            let b = before
                .iter()
                .find(|x| x.belongs_to_var == fact.belongs_to_var)
                .context("Variable mismatch.")?;

            edges.push(Edge::CallToReturn {
                from: b.clone().clone(),
                to: fact.clone(),
            });
        }

        Ok(edges)
    }

    fn tabulate(&mut self, mut graph: &mut Graph, prog: &Program, req: &Request) -> Result<()> {
        debug!("Convert intermediate repr to graph");

        let function = prog
            .functions
            .iter()
            .find(|x| x.name == req.function)
            .context("Cannot find function")?;

        if graph.is_function_defined(&function.name) {
            debug!("==> Function was already summarised.");
            return Ok(());
        }

        let facts = graph.init_function(&function)?;
        let init = facts.get(0).unwrap().clone();

        let mut path_edge = Vec::new();
        let mut worklist = VecDeque::new();
        let mut summary_edge = Vec::new();

        path_edge.push(Edge::Path {
            from: init.clone(),
            to: init.clone(),
        });
        worklist.push_back(Edge::Path {
            from: init.clone(),
            to: init.clone(),
        });

        self.forward(
            &prog,
            &function,
            &mut path_edge,
            &mut worklist,
            &mut summary_edge,
            &mut graph,
        )?;

        Ok(())
    }

    fn propagate(
        &self,
        path_edge: &mut Vec<Edge>,
        worklist: &mut VecDeque<Edge>,
        e: Edge,
    ) -> Result<()> {
        let f = path_edge.iter().rev().find(|x| *x == &e);

        if f.is_none() {
            path_edge.push(e.clone());
            worklist.push_back(e);
        }

        Ok(())
    }

    fn forward(
        &self,
        program: &Program,
        function: &AstFunction,
        path_edge: &mut Vec<Edge>,
        worklist: &mut VecDeque<Edge>,
        summary_edge: &mut Vec<Edge>,
        graph: &mut Graph,
    ) -> Result<()> {
        let mut end_summary: HashMap<(String, usize, String), Vec<Fact>> = HashMap::new();
        let mut incoming: HashMap<(String, usize, String), Vec<Fact>> = HashMap::new();

        while let Some(edge) = worklist.pop_front() {
            debug!("Popping edge from worklist {:#?}", edge);

            let d1 = edge.get_from();
            let d2 = edge.to();

            let pc = edge.to().pc;
            debug!("Instruction pointer is {}", pc);

            let instructions = &program
                .functions
                .iter()
                .find(|x| x.name == d2.function)
                .context("Cannot find function")?
                .instructions;
            let n = instructions.get(pc);
            debug!("=> Instruction {:?}", n);

            if let Some(n) = n {
                match n {
                    Instruction::Call(callee, params, dest) => {
                        let call_edges =
                            self.pass_args(program, function, callee, params, dest, graph, d2.pc)?;

                        for d3 in call_edges.into_iter() {
                            debug!("d3 {:?}", d3);

                            self.propagate(
                                path_edge,
                                worklist,
                                Edge::Path {
                                    from: d3.to().clone(),
                                    to: d3.to().clone(),
                                },
                            )?; //self loop

                            debug!(
                                "Propagate {:?}",
                                Edge::Path {
                                    from: d3.to().clone(),
                                    to: d3.to().clone(),
                                }
                            );

                            //Add incoming

                            if let Some(incoming) = incoming.get_mut(&(
                                d3.to().function.clone(),
                                d3.to().pc,
                                d3.to().belongs_to_var.clone(),
                            )) {
                                incoming.push(d2.clone());
                            } else {
                                incoming.insert(
                                    (
                                        d3.to().function.clone(),
                                        d3.to().pc,
                                        d3.to().belongs_to_var.clone(),
                                    ),
                                    vec![d2.clone()],
                                );
                            }

                            debug!("Incoming in call {:#?}", incoming);

                            if let Some(end_summary) = end_summary.get(&(
                                d3.to().function.clone(),
                                d3.to().pc,
                                d3.to().belongs_to_var.clone(),
                            )) {
                                for d4 in end_summary.iter() {
                                    for d5 in self.return_val(
                                        &function.name,
                                        &d4.function,
                                        d2.pc,
                                        d4.pc,
                                        &instructions,
                                        graph,
                                    )? {
                                        summary_edge.push(Edge::Summary {
                                            from: d2.clone(),
                                            to: d5.get_from().clone(), //return_site is d5.get_from?
                                        });
                                    }
                                }
                            }

                            debug!("End summary {:#?}", end_summary);
                        }

                        let call_flow =
                            self.call_flow(program, function, callee, params, dest, graph, pc)?;

                        let new_function = program
                            .functions
                            .iter()
                            .find(|x| x.name == d2.function)
                            .unwrap();

                        //TODO `or` with overwritten value
                        for d3 in call_flow {
                            let taut = graph.get_taut(&d3.get_from().function).unwrap().clone();
                            self.propagate(
                                path_edge,
                                worklist,
                                Edge::Path {
                                    from: taut,
                                    to: d3.to().clone(),
                                },
                            )?; // adding edges to return site of caller from d1
                        }
                    }
                    _ => {
                        let new_function = program
                            .functions
                            .iter()
                            .find(|x| x.name == d2.function)
                            .unwrap();
                        for f in self
                            .flow(&new_function, graph, d2.pc, &d2.belongs_to_var)?
                            .iter()
                        {
                            debug!("Normal flow {:?}", f);
                            let to = f.to();
                            self.propagate(
                                path_edge,
                                worklist,
                                Edge::Path {
                                    from: d1.clone(),
                                    to: to.clone(),
                                },
                            )?;
                        }
                    }
                }
            } else {
                // this is E_p
                debug!("=> Reached end of procedure");

                assert_eq!(d1.function, d2.function);

                // Summary
                if let Some(end_summary) =
                    end_summary.get_mut(&(d1.function.clone(), d1.pc, d1.belongs_to_var.clone()))
                {
                    let facts = graph
                        .get_facts_at(&d2.function.clone(), d2.pc)?
                        .into_iter()
                        .map(|x| x.clone());
                    end_summary.extend(facts);
                } else {
                    let facts = graph
                        .get_facts_at(&d2.function.clone(), d2.pc)?
                        .into_iter()
                        .map(|x| x.clone())
                        .collect();
                    end_summary.insert(
                        (d1.function.clone(), d1.pc, d1.belongs_to_var.clone()),
                        facts,
                    );
                }

                debug!("End Summary {:#?}", end_summary);

                // Incoming
                if let Some(incoming) =
                    incoming.get_mut(&(d1.function.clone(), d1.pc, d1.belongs_to_var.clone()))
                {
                    for d4 in incoming {
                        let instructions = &program
                            .functions
                            .iter()
                            .find(|x| x.name == d4.function)
                            .context("Cannot find function")?
                            .instructions;

                        let ret_vals = self.return_val(
                            &d4.function,
                            &d2.function,
                            d4.pc,
                            d2.pc,
                            &instructions,
                            graph,
                        )?;

                        let return_site_facts = graph
                            .get_facts_at(&d4.function, d4.pc + 1)?
                            .into_iter()
                            .map(|x| x.clone())
                            .collect::<Vec<_>>();

                        for calling_edge in ret_vals.into_iter() {
                            let return_site_fact_of_caller = return_site_facts
                                .iter()
                                .find(|x| {
                                    x.belongs_to_var == calling_edge.get_from().belongs_to_var
                                })
                                .context("Cannot find return site of caller")?;

                            if summary_edge
                                .iter()
                                .find(|x| {
                                    x.get_from() != d4 && x.to() != return_site_fact_of_caller
                                }) //d5 muss hier im test sein
                                .is_none()
                            {
                                summary_edge.push(Edge::Summary {
                                    from: d4.clone(),
                                    to: return_site_fact_of_caller.clone(),
                                });

                                let edges: Vec<_> = path_edge
                                    .iter()
                                    .filter(|x| {
                                        x.to() == d4 && &x.get_from().function == &d4.function
                                    })
                                    .cloned()
                                    .collect();

                                for d3 in edges.into_iter() {
                                    self.propagate(
                                        path_edge,
                                        worklist,
                                        Edge::Path {
                                            from: d3.get_from().clone(),
                                            to: return_site_fact_of_caller.clone(),
                                        },
                                    )?;
                                }
                            }
                        }
                    }
                }
            }
        }

        graph.edges.extend_from_slice(&path_edge);
        graph.edges.extend_from_slice(&summary_edge);

        Ok(())
    }
}
