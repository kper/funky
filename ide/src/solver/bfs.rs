use crate::icfg::graph2::Graph;
use crate::solver::*;
use std::collections::VecDeque;

use log::debug;

pub struct BFS;

impl GraphReachability for BFS {
    fn all_sinks(&mut self, graph: &mut Graph, req: Request) -> Vec<Taint> {
        let mut queue: VecDeque<usize> = VecDeque::new();
        let mut seen = Vec::new();

        if let Some(start) = graph.query(&req) {
            debug!("Adding start {:#?}", start);
            queue.push_back(start.id);
        } else {
            return Vec::new();
        }

        while let Some(node) = queue.pop_front() {
            debug!("Popping node {}", node);
            seen.push(node);
            for child in graph.get_neighbours(node) {
                debug!("Adding child {:#?}", graph.query_by_fact_id(child));

                if !seen.contains(&child) {
                    debug!("queue child {}", child);
                    queue.push_back(child);
                }
            }
        }

        seen.into_iter()
            .map(|x| graph.query_by_fact_id(x).unwrap())
            .map(|x| Taint {
                from: req.variable.clone(),
                from_function: req.function.clone(),
                from_pc: req.pc,
                to: x.belongs_to_var.clone(),
                to_pc: x.pc,
                to_function: x.function.clone(),
            })
            .collect()
    }
}
