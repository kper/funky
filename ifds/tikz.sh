#!/bin/bash

rm /tmp/graph.tex
rm /tmp/graph.pdf
tail -n +6 $1 > /tmp/graph.tex
xelatex /tmp/graph.tex && evince graph.pdf
