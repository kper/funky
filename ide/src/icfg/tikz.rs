use crate::icfg::graph2::{Edge, Graph};
use log::debug;

pub const TAU: usize = 1;

pub fn render_to(graph: &Graph) -> String {
    let mut str_vars = String::new();

    let mut index = 0;
    let mut functions: Vec<_> = graph.functions.iter().map(|(_, f)| f).collect();

    functions.sort_by(|a, b| b.name.cmp(&a.name));

    for function in functions {
        let function_name = &function.name;
        debug!("Drawing function {}", function_name);

        let facts = graph.facts.iter().filter(|x| &x.function == function_name);
        let notes = graph.notes.iter().filter(|x| &x.function == function_name);

        for fact in facts {
            debug!("Drawing fact");

            str_vars.push_str(&format!(
                "\\node[circle,fill,inner sep=1pt,label=left:{}] ({}) at ({}, {}) {{}};\n",
                fact.belongs_to_var.replace("%", "\\%"),
                fact.id,
                index + fact.track,
                fact.pc,
            ));
        }

        for note in notes {
            debug!("Drawing note");

            str_vars.push_str(&format!(
                "\\node[font=\\tiny] (note_{}) at ({}, {}) {{{}}};\n",
                note.id,
                index as f64 - 1.5,   //x
                note.pc as f64 - 0.5, //y
                note.note.replace("%", "").replace("\"", "").escape_debug(),
            ));
        }

        index += function.definitions + TAU + 1;
    }

    for edge in graph.edges.iter() {
        match edge {
            Edge::Normal { from, to , curved } => {
                if *curved {
                    str_vars.push_str(&format!("\t\t\\path[->, bend right] ({}) edge ({});\n", from.id, to.id));
                }
                else {
                    str_vars.push_str(&format!("\t\t\\path[->] ({}) edge ({});\n", from.id, to.id));
                }
            }
            Edge::Call { from, to } => {
                str_vars.push_str(&format!(
                    "\t\t\\path[->, green] ({}) [bend left] edge node {{ }} ({});\n",
                    from.id, to.id
                ));
            }
            Edge::CallToReturn { from, to } => {
                str_vars.push_str(&format!(
                    "\t\t\\path[->] ({}) edge node {{ }} ({});\n",
                    from.id, to.id
                ));
            }
            Edge::Return { from, to } => {
                str_vars.push_str(&format!(
                    "\t\t\\path[->, red] ({}) [bend right] edge  node {{ }} ({});\n",
                    from.id, to.id
                ));
            }
        }
    }

    template(str_vars)
}

fn template(inject: String) -> String {
    format!(
        "   \\documentclass{{article}}
            \\usepackage{{pgf, tikz}}
            \\usetikzlibrary{{arrows, automata}}
            \\begin{{document}}
                \\begin{{tikzpicture}}
                    [node distance=3cm]
                    \\tikzstyle{{every state}}=[
                        draw = black,
                        thick,
                        fill = white,
                        minimum size = 4mm
                    ]

                    {}

                \\end{{tikzpicture}}
            \\end{{document}}",
        inject
    )
}
