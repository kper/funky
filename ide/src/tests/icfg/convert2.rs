use std::env::var;

use crate::icfg::convert2::ConvertSummary;
use crate::icfg::tikz::render_to;
use crate::solver::Request;
use insta::assert_snapshot;

use crate::grammar::*;

macro_rules! ir {
    ($name:expr, $req:expr, $ir:expr) => {
        let mut convert = ConvertSummary::new();

        let prog = ProgramParser::new().parse(&$ir).unwrap();

        let graph = convert.visit(&prog, &$req).unwrap();

        let output = render_to(&graph);

        assert_snapshot!(format!("{}_dot", $name), output);
    };
}

#[test]
fn test_ir_const() {
    let req = Request {
        variable: "%0".to_string(),
        function: "test".to_string(),
        pc: 1,
    };

    ir!(
        "test_ir_const",
        req,
        "
         define test (result 0) (define %0) {
            %0 = 1
         };
    "
    );
}

#[test]
fn test_ir_double_assign() {
    let req = Request {
        variable: "%2".to_string(),
        function: "test".to_string(),
        pc: 3,
    };
    ir!(
        "test_ir_double_const",
        req,
        "
         define test (result 0) (define %0 %1 %2){
            %0 = 1
            %1 = 1
            %2 = %0
            %2 = %1
         };
    "
    );
}

#[test]
fn test_ir_unop() {
    let req = Request {
        variable: "%0".to_string(),
        function: "test".to_string(),
        pc: 1,
    };
    ir!(
        "test_ir_unop",
        req,
        "define test (result 0) (define %0 %1) {
            %0 = 1
            %1 = op %0
            %1 = op %0   
        };"
    );
}

#[test]
fn test_ir_binop() {
    let req = Request {
        variable: "%0".to_string(),
        function: "test".to_string(),
        pc: 1,
    };
    ir!(
        "test_ir_binop",
        req,
        "define test (result 0) (define %0 %1 %2) {
            %0 = 1
            %1 = 1
            %2 = %0 op %1
            %2 = %1 op %0   
        };"
    );
}

#[test]
fn test_ir_killing_op() {
    let req = Request {
        variable: "%0".to_string(),
        function: "test".to_string(),
        pc: 1,
    };
    ir!(
        "test_ir_killing_op",
        req,
        "define test (result 0) (define %0 %1 %2)  {
            %0 = 1
            %1 = 1
            KILL %0
            KILL %1
            %2 = 1
        };"
    );
}

#[test]
fn test_ir_block() {
    let req = Request {
        variable: "%0".to_string(),
        function: "test".to_string(),
        pc: 2,
    };
    ir!(
        "test_ir_block",
        req,
        "define test (result 0) (define %0 %1) {
            BLOCK 0
            %0 = 1
            GOTO 1
            BLOCK 1
            %1 = 2
        };"
    );
}

#[test]
fn test_ir_if_else() {
    let req = Request {
        variable: "%0".to_string(),
        function: "main".to_string(),
        pc: 2,
    };
    ir!(
        "test_ir_if_else",
        req,
        "define main (result 0) (define %0 %1 %2) {
            BLOCK 0
            %0 = 1
            IF %1 THEN GOTO 1 ELSE GOTO 2 
            BLOCK 1
            %1 = 2
            %2 = 3
            GOTO 3
            BLOCK 2
            %2 = 4
            GOTO 3
            BLOCK 3
            %0 = %2
        };
        "
    );
}

#[test]
fn test_ir_if() {
    let req = Request {
        variable: "%0".to_string(),
        function: "main".to_string(),
        pc: 2,
    };
    ir!(
        "test_ir_if",
        req,
        "define main (result 0) (define %0 %1 %2) {
            BLOCK 0
            %0 = 1
            IF %1 THEN GOTO 0
            %1 = 2
            %2 = 3
        };
        "
    );
}

#[test]
fn test_ir_loop() {
    let req = Request {
        variable: "%0".to_string(),
        function: "main".to_string(),
        pc: 2,
    };
    ir!(
        "test_ir_loop",
        req,
        "define main (result 0) (define %0 %1) {
            BLOCK 0
            %0 = 1
            GOTO 0 
        };
        "
    );
}

#[test]
fn test_ir_table() {
    let req = Request {
        variable: "%0".to_string(),
        function: "main".to_string(),
        pc: 2,
    };
    ir!(
        "test_ir_table",
        req,
        "define main (result 0) (define %0 %1 %2) {
            BLOCK 0
            %0 = 1
            %1 = 2
            %2 = 3
            BLOCK 1
            %1 = 2
            %2 = 3
            BLOCK 2
            %2 = 4
            TABLE GOTO 0 1 2 ELSE GOTO 2
        };
        "
    );
}

#[test]
fn test_ir_functions() {
    let req = Request {
        variable: "%0".to_string(),
        function: "test".to_string(),
        pc: 1,
    };
    ir!(
        "test_ir_functions",
        req,
        "define test (result 0) (define %0) {
            %0 = 1
            CALL mytest(%0)
        };
        define mytest (param %0) (result 0) (define %0 %1)  {
            %0 = 2   
            %1 = 3
            RETURN;
        };"
    );
}

#[test]
fn test_ir_return_values() {
    let req = Request {
        variable: "%0".to_string(),
        function: "test".to_string(),
        pc: 1,
    };
    ir!(
        "test_ir_return_values",
        req,
        "define test (result 0) (define %0 %1) {
            %0 = 1
            %1 <- CALL mytest(%0)
        };
        define mytest (param %0) (result 1) (define %0 %1) {
            %0 = 2   
            %1 = 3
            RETURN %1;
        };"
    );
}

#[test]
fn test_ir_return_values2() {
    let req = Request {
        variable: "%0".to_string(),
        function: "test".to_string(),
        pc: 1,
    };
    ir!(
        "test_ir_return_values2",
        req,
        "define test (result 0) (define %0 %1 %2) {
            %0 = 1
            %1 = 2
            %2 <- CALL mytest(%0)
        };
        define mytest (param %0) (result 1) (define %0 %1) {
            %1 = 3
            RETURN %0;
        };"
    );
}

#[test]
fn test_ir_return_values3() {
    let req = Request {
        variable: "%0".to_string(),
        function: "test".to_string(),
        pc: 1,
    };
    ir!(
        "test_ir_return_values3",
        req,
        "define test (result 0) (define %0 %1 %2) {
            %0 = 1
            %1 = 2
            %2 <- CALL mytest(%0)
        };
        define mytest (param %0) (result 1) (define %0 %1) {
            %1 = 3
            RETURN %1;
        };"
    );
}

#[test]
fn test_ir_overwrite_return_values() {
    let req = Request {
        variable: "%0".to_string(),
        function: "test".to_string(),
        pc: 1,
    };
    ir!(
        "test_ir_overwrite_return_values",
        req,
        "define test (result 0) (define %0 %1) {
            %0 = 1
            %0 <- CALL mytest(%0)
        };
        define mytest (param %0) (result 1) (define %0 %1) {
            %0 = 2   
            %1 = 3
            RETURN %1;
        };"
    );
}

#[test]
fn test_ir_early_return() {
    let req = Request {
        variable: "%0".to_string(),
        function: "test".to_string(),
        pc: 1,
    };
    ir!(
        "test_ir_early_return",
        req,
        "define test (result 0) (define %0 %1 %2) {
            %0 = 1
            %1 <- CALL mytest(%0)
            %2 <- CALL mytestfoo(%0)
        };
        define mytest (param %0) (result 1) (define %0 %1) {
            %0 = 2   
            RETURN %0;
            %1 = 3
            RETURN %1;
        };
        define mytestfoo (param %0) (result 1) (define %0 %1) {
            %0 = 2
            %1 = 3
            RETURN %0;
        };
        "
    );
}

#[test]
fn test_ir_return() {
    let req = Request {
        variable: "%0".to_string(),
        function: "test".to_string(),
        pc: 1,
    };
    ir!(
        "test_ir_return",
        req,
        "define test (result 0) (define %0 %1) {
            %0 = 1
            %0 %1 <- CALL mytest(%0)
            %1 = 2
        };
        define mytest (param %0) (result 2) (define %0 %1) {
            %0 = 2   
            %1 = 3
            RETURN %0 %1;
        };
        "
    );
}
