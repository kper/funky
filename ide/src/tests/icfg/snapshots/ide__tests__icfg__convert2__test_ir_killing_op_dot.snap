---
source: src/tests/icfg/convert2.rs
expression: output

---
   \documentclass{standalone}
            \usepackage{pgf, tikz}
            \usetikzlibrary{arrows, automata}
            \begin{document}
                \begin{tikzpicture}
                    [node distance=3cm]
                    \tikzstyle{every state}=[
                        draw = black,
                        thick,
                        fill = white,
                        minimum size = 4mm
                    ]

                    \node[circle,fill,inner sep=1pt,label=left:taut] (20) at (0, 5) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (16) at (0, 4) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (12) at (0, 3) { };
\node[circle,fill,inner sep=1pt,label=left:\%0] (9) at (1, 2) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (8) at (0, 2) { };
\node[circle,fill,inner sep=1pt,label=left:\%0] (5) at (1, 1) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (4) at (0, 1) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (0) at (0, 0) { };
\node[font=\tiny] (note_0) at (-1.5, 0.5) {Const(0, 1.0)};
\node[font=\tiny] (note_1) at (-1.5, 1.5) {Const(1, 1.0)};
\node[font=\tiny] (note_2) at (-1.5, 2.5) {Kill(0)};
\node[font=\tiny] (note_3) at (-1.5, 3.5) {Kill(1)};
\node[font=\tiny] (note_4) at (-1.5, 4.5) {Const(2, 1.0)};
		\path[->] (0) edge (4);
		\path[->] (5) edge (9);
		\path[->] (5) edge (9);
		\path[->] (4) edge (8);
		\path[->] (8) edge (12);
		\path[->] (12) edge (16);
		\path[->] (16) edge (20);
		\path[->, blue] (0) [loop right] edge  node { } (0);
		\path[->] (0) edge (5);
		\path[->, blue] (0) [bend right] edge  node { } (5);
		\path[->, blue] (0) [bend right] edge  node { } (4);
		\path[->, blue] (0) [bend right] edge  node { } (9);
		\path[->, blue] (0) [bend right] edge  node { } (8);
		\path[->, blue] (0) [bend right] edge  node { } (12);
		\path[->, blue] (0) [bend right] edge  node { } (16);
		\path[->, blue] (0) [bend right] edge  node { } (20);


                \end{tikzpicture}
            \end{document}
